mat64/svd.go

```diff
@@ -6,474 +6,165 @@
-	// in sigma and the super-diagonal elements in e.
-	nct := min(m-1, n)
-	nrt := max(0, min(n-2, m))
-	for k := 0; k < max(nct, nrt); k++ {
-		if k < nct {
-			// Compute the transformation for the k-th column and
-			// place the k-th diagonal in sigma[k].
-			// Compute 2-norm of k-th column without under/overflow.
-			sigma[k] = 0
-			for i := k; i < m; i++ {
-				sigma[k] = math.Hypot(sigma[k], a.at(i, k))
-			}
-			if sigma[k] != 0 {
-				if a.at(k, k) < 0 {
-					sigma[k] = -sigma[k]
-				}
-				for i := k; i < m; i++ {
-					a.set(i, k, a.at(i, k)/sigma[k])
-				}
-				a.set(k, k, a.at(k, k)+1)
-			}
-			sigma[k] = -sigma[k]
+	var jobU, jobVT lapack.SVDJob
+	switch kind {
+	default:
+		panic("svd: bad input kind")
+	case matrix.SVDNone:
+		jobU = lapack.SVDNone
+		jobVT = lapack.SVDNone
+	case matrix.SVDFull:
+		svd.u = blas64.General{
+			Rows:   m,
+			Cols:   m,
+			Stride: m,
+			Data:   make([]float64, m*m),
 		}
-
-		for j := k + 1; j < n; j++ {
-			if k < nct && sigma[k] != 0 {
-				// Apply the transformation.
-				var t float64
-				for i := k; i < m; i++ {
-					t += a.at(i, k) * a.at(i, j)
-				}
-				t = -t / a.at(k, k)
-				for i := k; i < m; i++ {
-					a.set(i, j, a.at(i, j)+t*a.at(i, k))
-				}
-			}
-
-			// Place the k-th row of a into e for the
-			// subsequent calculation of the row transformation.
-			e[j] = a.at(k, j)
+		svd.vt = blas64.General{
+			Rows:   n,
+			Cols:   n,
+			Stride: n,
+			Data:   make([]float64, n*n),
 		}
-
-		if wantu && k < nct {
-			// Place the transformation in u for subsequent back
-			// multiplication.
-			for i := k; i < m; i++ {
-				u.set(i, k, a.at(i, k))
-			}
+		jobU = lapack.SVDAll
+		jobVT = lapack.SVDAll
+	case matrix.SVDThin:
+		svd.u = blas64.General{
+			Rows:   m,
+			Cols:   min(m, n),
+			Stride: min(m, n),
+			Data:   make([]float64, m*min(m, n)),
 		}
-
-		if k < nrt {
-			// Compute the k-th row transformation and place the
-			// k-th super-diagonal in e[k].
-			// Compute 2-norm without under/overflow.
-			e[k] = 0
-			for i := k + 1; i < n; i++ {
-				e[k] = math.Hypot(e[k], e[i])
-			}
-			if e[k] != 0 {
-				if e[k+1] < 0 {
-					e[k] = -e[k]
-				}
-				for i := k + 1; i < n; i++ {
-					e[i] /= e[k]
-				}
-				e[k+1]++
-			}
-			e[k] = -e[k]
-			if k+1 < m && e[k] != 0 {
-				// Apply the transformation.
-				for i := k + 1; i < m; i++ {
-					work[i] = 0
-				}
-				for j := k + 1; j < n; j++ {
-					for i := k + 1; i < m; i++ {
-						work[i] += e[j] * a.at(i, j)
-					}
-				}
-				for j := k + 1; j < n; j++ {
-					t := -e[j] / e[k+1]
-					for i := k + 1; i < m; i++ {
-						a.set(i, j, a.at(i, j)+t*work[i])
-					}
-				}
-			}
-			if wantv {
-				// Place the transformation in v for subsequent
-				// back multiplication.
-				for i := k + 1; i < n; i++ {
-					v.set(i, k, e[i])
-				}
-			}
+		svd.vt = blas64.General{
+			Rows:   min(m, n),
+			Cols:   n,
+			Stride: n,
+			Data:   make([]float64, min(m, n)*n),
 		}
+		jobU = lapack.SVDInPlace
```

Comment:
Yes, they probably should. It's more memory efficient to have one of them be written in place into A (the larger one). TODO added.